### PR TITLE
Prevent run_onbuild in editable_mode and VersioningitSdist for PEP 660

### PR DESCRIPTION
**Pull request by basnijholt**
I believe this will fix the issue I mentioned in https://github.com/pypa/setuptools/issues/3501#issuecomment-1212393121.

edit: @jwodder, I have just tested this and it works :tada:

### CODE REVIEW

**Comment by jwodder** [comment]
@basnijholt In order to prevent regressions in the future, can you provide or describe a test case that fails under the current code and succeeds with this patch?  The current test suite passes without a problem, so it never caught whatever went wrong for you.  More importantly, I'd also like to know exactly what is going wrong, as I'm still not clear on it.

**Comment by basnijholt** [comment]
Unfortunately, I will not have time to add a test in the coming week. It would be great to have this fix in a release before I will be able to make time though.

**Comment by jwodder** [comment]
@basnijholt Could you at least describe a test case — e.g., by linking to a repository that exhibits the bug and including the command you ran that failed?

**Comment by basnijholt** [comment]
The repository is a private one. I can try to cook up an example next week 👍

The error was the same as in the issue I linked in the initial post.

**Comment by jwodder** [comment]
@basnijholt The only part of that post that seems relevant is the error message; everything else there is specific to versioneer.  Moreover, when I try running `pip install -e .` myself on a versioningit project that uses onbuild, it works fine.  I therefore remain unclear on exactly what the problem you're experiencing with versioningit is.

**Comment by basnijholt** [comment]
@jwodder, I have made an example here https://github.com/basnijholt/versioningit_fail

just
```
git clone git@github.com:basnijholt/versioningit_fail.git
cd versioningit_fail
pip install -v -e .
```
this will now work (because https://github.com/pypa/setuptools/issues/3501 is fixed), however, it will print a big warning:
```
  running editable_wheel
  creating /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info
  writing /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/PKG-INFO
  writing dependency_links to /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/dependency_links.txt
  writing requirements to /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/requires.txt
  writing top-level names to /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/top_level.txt
  writing manifest file '/private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/SOURCES.txt'
  reading manifest file '/private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/SOURCES.txt'
  writing manifest file '/private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example.egg-info/SOURCES.txt'
  creating '/private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example-0.0.dist-info'
  creating /private/tmp/pip-wheel-baccd2sf/tmpohdn32uz/example-0.0.dist-info/WHEEL
  /private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
    warnings.warn(
  running build_py
  /private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/command/editable_wheel.py:283: SetuptoolsDeprecationWarning: Traceback (most recent call last):
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/command/editable_wheel.py", line 289, in _safely_run
      return self.run_command(cmd_name)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
      self.distribution.run_command(command)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 992, in run_command
      cmd_obj.run()
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/versioningit/cmdclasses.py", line 61, in run
      run_onbuild(
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/versioningit/core.py", line 412, in run_onbuild
      vgit.do_onbuild(build_dir=build_dir, is_source=is_source, version=version)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/versioningit/core.py", line 262, in do_onbuild
      self.onbuild(build_dir=build_dir, is_source=is_source, version=version)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/versioningit/methods.py", line 154, in __call__
      return self.method(params=self.params, **kwargs)
    File "/private/tmp/pip-build-env-qujdfdys/overlay/lib/python3.9/site-packages/versioningit/onbuild.py", line 62, in replace_version_onbuild
      lines = path.read_text(encoding=encoding).splitlines(keepends=True)
    File "/Users/basnijholt/mambaforge/lib/python3.9/pathlib.py", line 1266, in read_text
      with self.open(mode='r', encoding=encoding, errors=errors) as f:
    File "/Users/basnijholt/mambaforge/lib/python3.9/pathlib.py", line 1252, in open
      return io.open(self, mode, buffering, encoding, errors, newline,
    File "/Users/basnijholt/mambaforge/lib/python3.9/pathlib.py", line 1120, in _opener
      return self._accessor.open(self, flags, mode)
  FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmphnnaigov.build-lib/example/_version.py'


              If you are seeing this warning it is very likely that a setuptools
              plugin or customization overrides the `build_py` command, without
              tacking into consideration how editable installs run build steps
              starting from v64.0.0.

              Plugin authors and developers relying on custom build steps are encouraged
              to update their `build_py` implementation considering the information in
              https://setuptools.pypa.io/en/latest/userguide/extension.html
              about editable installs.

              For the time being `setuptools` will silence this error and ignore
              the faulty command, but this behaviour will change in future versions.
```

**Comment by abravalheri** [comment]
Hi @basnijholt, I guess that what is happening is the something similar to following:

- `_version.py` is dynamically generated in the `build_lib` or `build_temp` directory
- `editable_wheel` uses a `build_lib`/`build_temp` outside of the project root
- somehow `setuptools` tries to consider this file outside the project root as part of `package_data` or `data_files`.
  - I don't know if this was already happened before, but since `build_lib`/`build_temp` were inside the project root, setuptools would chill about it *OR*
  - This happens because now `setuptools` will try harder to find "source files" (this was a change introduced in the new release).

Regardless, I add some filtering-out in pypa/setuptools#3517 that might solve this problem.

But yeah, files created in the external `build_lib` and `build_temp` would probably just be ignored (unless they are listed in [`get_outputs()` and `get_output_mapping()`](https://setuptools.pypa.io/en/latest/userguide/extension.html#supporting-sdists-and-editable-installs-in-build-sub-commands)), so a developer might decide to optimize it and not generate them in the first place.

TL;DR: Let's see if the problem persists after the next release.

**Comment by jwodder** [comment]
@basnijholt I indeed see the warning when building that example repo with the given command.  However, the `pyproject.toml` file in that repo specifies `versioningit ~= 1.0`, which means that you'll be limited to versions less than 2.0 and so won't receive the PEP 660 patch in v2.0.1.  When I change that to `versioningit ~= 2.0` and run the command using versioningit 2.0.1, I don't get the warning or any failures.  Can you confirm this?

**Comment by basnijholt** [comment]
Wow, I don't know how I missed this.

Locally I tested it like `"versioningit @ file:///Users/basnijholt/Downloads/versioningit"` where I switched between `master` and my patched branch, and I could reproduce the problem. Now I indeed see that setting `versioningit ~= 2.0` just works. I am not sure whether I just broke my env or something else.

My apologies for the noise :-)

**Comment by jwodder** [comment]
@basnijholt No need to apologize.  I think I'll release a backport of the 2.0.1 fix for the 1.x line for any projects still using versioningit ~= 1.0.  (The v0.x sequence shouldn't need any fixes, as the "onbuild" step didn't exist then.)
